core/async/src/examples/actix_component_test.rs

```diff
@@ -0,0 +1,102 @@
+                (request.callback)(Ok(response));
+            }
+        },
+    )
+}
+
+#[test]
+fn test_actix_component() {
+    let builder = TestLoopBuilder::<ExampleComponentTestEvent>::new();
+    let sender = builder.sender();
+    let mut test = builder.build(ExampleComponentTestData {
+        dummy: (),
+        example: ExampleComponent::new(sender.clone().into_sender()),
+        outer: OuterComponent::new(
+            // We need to specify the explicit type here because this is a
+            // three-step conversion. First, we're converting the
+            // `DelaySender<ExampleComponentTestEvent>` into a
+            // `Sender<ExampleComponentAdapterMessage>`, which is a single
+            // sender that accepts an enum (generated by MultiSendMessage)
+            // that contains every type of message that can be sent to
+            // ExampleComponent. Then, we break it apart so that it becomes
+            // something that can send any of the messages that the enum
+            // contains, and finally we convert it a multisender
+            // ExampleComponentAdapter.
+            IntoSender::<ExampleComponentAdapterMessage>::into_sender(sender)
+                .break_apart()
+                .into_multi_sender(),
+        ),
+        periodic_requests_captured: vec![],
+    });
+    // This is to allow futures to be used in the test even though the
+    // test itself is synchronous.
+    test.register_handler(drive_futures().widen());
+    // This is to allow the ExampleComponent to run delayed actions (timers).
+    test.register_handler(drive_delayed_action_runners::<ExampleComponent>().widen());
+    // This is to capture the periodic requests sent by the ExampleComponent
+    // so we can assert against it.
+    test.register_handler(capture_events::<PeriodicRequest>().widen());
+    // This is to handle the ExampleComponentAdapterMessage events by
+    // forwarding them to the ExampleComponent.
+    test.register_handler(example_handler().widen());
+
+    // We need to redo whatever the ExampleActor does in its `started` method.
+    test.data.example.start(&mut test.delayed_action_runner());
+    // Send some requests; this can be done in the asynchronous context.
+    test.future_spawner().spawn("wait for 5", {
+        let res = test.data.outer.call_example_component_for_response(5);
+        async move {
+            assert_eq!(res.await, 5);
+        }
+    });
+    test.future_spawner().spawn("wait for 6", {
+        let res = test.data.outer.call_example_component_for_response(6);
+        async move {
+            assert_eq!(res.await, 6);
+        }
+    });
+    // Run for 3 seconds (not real time, but in the test loop time).
+    // It should result in sending 3 periodic requests.
+    test.run_for(Duration::seconds(3));
+    assert_eq!(
+        test.data.periodic_requests_captured,
+        vec![PeriodicRequest { id: 0 }, PeriodicRequest { id: 1 }, PeriodicRequest { id: 2 },]
```

Comment:
I've spent 20 minutes trying to figure out how exactly periodic requests are pushed to `periodic_requests_captured`, but I failed miserably 😞 
could you please describe how that happens exactly

Reply:
So during the periodic timer (scheduled via the `DelayedActionRunner` which goes to the `TestLoopDelayedActionRunner` which sends a `TestLoopDelayedActionEvent` to the test loop, which is then handled via the `drive_delayed_action_runners` listener), the `ExampleComponent` uses the `Sender<PeriodicRequest>` it is given to send the `PeriodicRequest`. That `Sender<PeriodicRequest>` is provided by taking the DelaySender from the test loop; therefore when sending it, it becomes a `ExampleComponentTestEvent::PeriodicRequest(..)` and is scheduled onto the TestLoop. When it comes time to handle that event, the event listener responsible is `capture_events<PeriodicRequest>()` (widened to handle the whole `ExampleComponentTestEvent` enum). That event listener is based on a `Data` type of `Vec<PeriodicRequest>`, and it handles the event by writing the event to that vector. After widening, it writes into anything that can be `AsMut`-ed into a `Vec<PeriodicRequest>`, so that goes into the `Vec<PeriodicRequest>` included in `ExampleComponentTestData`.

Does that make sense?

Reply:
Weeeell it makes sense, but I agree that this mechanism must be documented somewhere.
You could point out to core/async/src/examples/sum_numbers_test.rs.
These links can be aggregated in core/async/src/examples/README.

Reply:
@robin-near thanks for the detailed explanation!

> After widening, it writes into anything that can be AsMut-ed into a Vec<PeriodicRequest>

this was the missing bit, now it all makes sense.